modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RowVersion.java

```diff
@@ -0,0 +1,137 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.storage.pagememory.mv;
+
+import java.nio.ByteBuffer;
+import org.apache.ignite.internal.pagememory.StorableBase;
+import org.apache.ignite.internal.pagememory.io.AbstractDataPageIo;
+import org.apache.ignite.internal.pagememory.io.IoVersions;
+import org.apache.ignite.internal.storage.pagememory.mv.io.RowVersionDataIo;
+import org.apache.ignite.internal.tx.Timestamp;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents row version inside row version chain.
+ */
+public class RowVersion extends StorableBase {
+    /**
+     * A 'timestamp' representing absense of a timestamp.
+     */
+    public static final Timestamp NULL_TIMESTAMP = new Timestamp(Long.MIN_VALUE, Long.MIN_VALUE);
+    /**
+     * Represents an absent partitionless link.
+     */
+    public static final long NULL_LINK = 0;
+
+    @Nullable
+    private final Timestamp timestamp;
+    private final long nextLink;
+    private final int valueSize;
+    private final ByteBuffer value;
+
+    /**
+     * Constructor.
+     */
+    public RowVersion(int partitionId, long nextLink, ByteBuffer value) {
+        this(partitionId, 0, null, nextLink, value);
+    }
+
+    /**
+     * Constructor.
+     */
+    public RowVersion(int partitionId, long link, @Nullable Timestamp timestamp, long nextLink, ByteBuffer value) {
+        super(partitionId);
+        link(link);
+
+        assert !NULL_TIMESTAMP.equals(timestamp) : "Null timestamp provided";
+
+        this.timestamp = timestamp;
+        this.nextLink = nextLink;
+        this.valueSize = value.limit();
+        this.value = value;
+    }
+
+    @Nullable
+    public Timestamp timestamp() {
+        return timestamp;
+    }
+
+    public Timestamp timestampForStorage() {
+        return timestamp == null ? NULL_TIMESTAMP : timestamp;
+    }
+
+    /**
+     * Returns partitionless link of the next version or {@code 0} if this version is the last in the chain (i.e. it's the oldest version).
+     *
+     * @return partitionless link of the next version or {@code 0} if this version is the last in the chain
+     */
+    public long nextLink() {
+        return nextLink;
+    }
+
+    public int valueSize() {
+        return valueSize;
+    }
+
+    public ByteBuffer value() {
+        return value;
+    }
+
+    public boolean hasNextLink() {
+        return nextLink != NULL_LINK;
+    }
+
+    boolean isTombstone() {
+        return valueSize() == 0;
+    }
+
+    boolean isUncommitted() {
+        return timestamp == null;
+    }
+
+    boolean isCommitted() {
+        return timestamp != null;
+    }
+
+    @Override
+    public int size() {
+        return timestampStoreSize() + nextLinkStoreSize() + valueStoreSize();
+    }
+
+    private int timestampStoreSize() {
```

Comment:
Can these be constants instead of methods? It's very unusual for Ignite to see code like this.

Reply:
I introduced the methods solely as an aid to the reader. When you see '4 + 8 + 8', you have to somehow lookup the information (by looking at other code) that will help you see WHY it's '4 + 8 + 8'. With the methods in place, it's self-explanatory, you don't need to read other code, you spend less time understanding the `size()` method.

I know that in Ignite people don't do that, and it's a pity. My opinion is that this needs to be improved. The performance should not suffer as the JIT compiler will easily inline it all, but the readability becomes a bit better.

Reply:
Please disregard the previous comment, I re-read your comment and understood that I was fighting windmills. I changed it to constants (where it's possible).